### ISSUE TITLE
Remove tools mode?

### ISSUE THREAD

**Issue by digama0** [metamath/metamath-exe#31]
One fairly large component in metamath.exe is "tools mode", which is a set of text editing utilities borrowed from BASIC reminiscent of unix command line operations like `sed`. Personally, I think that this is best done with another tool rather than bundling it in with the proof assistant and other metamath-specific functionality in metamath.exe. What do people think about removing it? Does metamath tools mode appear in anyone's workflow (or the CI), and if so do we have adequate replacement utilities?

**Comment by digama0** [comment]
This one in particular looks like a maintenance dead weight:
```
MM> HELP UPDATE
This command tags edits made to a program source.  The idea is to keep
all past history of a file in the file itself, in the form of comments.
UPDATE was written for a proprietary language that allowed nested C-style
comments, and it may not be generally useful without some modification.
Essentially a (Unix) diff-like algorithm looks for changes between an
original and a revised file and puts the original lines into the revised
file in the form of comments.  Currently it is not well documented and it
may be easiest just to type UPDATE <return> and answer the questions.
Try it on an original and edited version of a test file to see if you
find it useful.
Syntax:  UPDATE <originfile> <editedinfile> <editedoutfile> <tag> <match>
```

**Comment by wlammen** [comment]
FYI: a quick overview over the range of functions in TOOLS mode.

MM> tools
Entering the Text Tools utilities.  Type HELP for help, EXIT to exit.
TOOLS> help
This utility assists with some common file manipulations.
Most commands will perform an identical operation on each line of a file.
Use HELP ? to see list of help topics.
Note:  When an output file is created, any previous version is renamed,
with ~1 appended, and any ~1 renamed to ~2, etc. (up to ~9, which is lost).
Note:  All string-matching command arguments are case-sensitive.

Line-by-line editing commands:
  ADD - Add a specified string to each line in a file
  CLEAN - Trim spaces and tabs on each line in a file; convert characters
  DELETE - Delete a section of each line in a file
  INSERT - Insert a string at a specified column in each line of a file
  SUBSTITUTE - Make a simple substitution on each line of the file
  TAG - Like ADD, but restricted to a range of lines
  SWAP - Swap the two halves of each line in a file
Other file processing commands:
  BREAK - Break up (parse) a file into a list of tokens (one per line)
  BUILD - Build a file with multiple tokens per line from a list
  COUNT - Count the occurrences in a file of a specified string
  NUMBER - Create a list of numbers
  PARALLEL - Put two files in parallel
  REVERSE - Reverse the order of the lines in a file
  RIGHT - Right-justify lines in a file (useful before sorting numbers)
  SORT - Sort the lines in a file with key starting at specified string
  MATCH - Extract lines containing (or not) a specified string
  UNDUPLICATE - Eliminate duplicate occurrences of lines in a file
  DUPLICATE - Extract first occurrence of any line occurring more than
      once in a file, discarding lines occurring exactly once
  UNIQUE - Extract lines occurring exactly once in a file
  (UNDUPLICATE, DUPLICATE, and UNIQUE also sort the lines as a side effect.)
  UPDATE (deprecated) - Update a C program for revision control
  TYPE (10 lines) - Display 10 lines of a file; similar to Unix "head"
  COPY - Similar to Unix "cat" but safe (same input & output name allowed)
  SUBMIT - Run a script containing Tools commands.

Command syntax ([] means optional):
  From TOOLS prompt:  TOOLS> <command> [<arg1> <arg2>...]

**Comment by wlammen** [comment]
I never used any of this.

**Comment by wlammen** [comment]
I assume Norm has used these to update his website on us2.metamath.org.

**Comment by benjub** [comment]
I actually use TOOLS quite regularly, for instance each time I'm using "scripts/min.cmd" (and the script itself uses it).  I find it useful, and the integration with metamath is obviously seamless.  I know there are much more powerful standard command line tools, like grep, sed, and piping them, but they are "too" powerful, with many options, esoteric regexp syntax, huge man pages (compared to the well-written help for each TOOLS command), etc.  I know I should learn them eventually, but for the occasional user, TOOLS is very simple since it was designed to do the kind of tasks one often finds oneself wanting to do with mm files or associated logs/output of the metamath program, and only these.

If TOOLS is a fairly well separated part of the metamath program, it may not be too much of a burden to keep it ?

**Comment by digama0** [comment]
@benjub Can you be more specific about which tools you use, and how? There is a lot of arcana in the tools implementation, including code that looks like it was ported straight from an assembly language (or basic, it's hard to tell the difference sometimes). What do you mean by "integration with metamath"? As far as I know it may as well be a separate application (and indeed, that is another way to approach this issue, splitting off tools-mode into a separate executable and/or repo entirely).

**Comment by digama0** [comment]
I have no doubt that Norm used tools mode regularly in his workflow. It strikes me as something written specifically for this purpose, and it should come as no surprise since that's pretty much what it is. Not to put too fine a point on it, but if it's only helpful for Norm, then there isn't much use keeping it around and updated since he's not going to be using it anymore.

**Comment by digama0** [comment]
I took a look at [min.cmd](https://github.com/metamath/set.mm/blob/develop/scripts/min.cmd), and the spec looks bizarre. Can someone who uses this script explain what they want it to do? I think we should just support whatever it is more directly, rather than layering a weird custom scripting language on top of the metamath repl.

**Comment by benjub** [comment]
I typically use min.cmd to try and minimize theorems in list A with theorems/axioms in list B.  I follow the help written in the comments of min.cmd.  Namely, A is typically the (reformatted) output of `MM>show usage <label-match>` and B is actually the label-match which I write in the command `TOOLS> substitute 1.tmp xxx <label-match> ""`.

All in all, this uses the TOOLS commands: substitute, match, break, unduplicate, add. But I think some other commands are useful as well, like sort, clean, delete, insert.  Recently, I used a TOOLS command (paralell, I think) to make a bulk addition to the list of "recent changes" in set.mm.  I also use "match" to select lines from outputs of the min.cmd script that contain the string "decreased", and I often post the resulting lines in PR comments when I do some bulk minimizations.

**Comment by benjub** [comment]
By "integration in metamath", I mean that I can do all this while staying in the metamath program, that is, I type everyting at the prompt MM> (or MM-PA> or TOOLS>).

**Comment by digama0** [comment]
I think there are a couple things that can be done here:

1. Metamath should support better command-line usage for consumption by scripts, compatible with a unix-like bash script pipeline (which includes usage by tools mode). For example, it should be possible to run `./metamath -e set.mm "show usage foo"` and get the output from the `show usage` command in machine readable form (let's say, no `MM>` or echoing commands, no `press S to continue`, and one label per line).

2. We should have a cheat sheet for mapping tools mode to unix commands - in most cases there is a one-liner that has the same effect, and for the rest we can have a python script or similar to do the job (or delete if no one uses it)

3.  Tools mode can be split into a separate application. There are two ways to do this:
    * Clone metamath.exe and extract the tools stuff, to ensure exact compatibility.
    * Rewrite it in rust or python (which can do it with a lot less fuss). Compatibility will be best effort, with a focus on the most-used commands.

    In either case, the main program can replace `MM> TOOLS` to call `mm-tools` as a sub-shell, or alternatively we can generalize it to `MM> RUN mm-tools` so that people can `run bash` too.

(3) offers the best opportunity for migration without changing observable behavior, for those that can't break their muscle memory. I think (1) and (2) is a better way forward so that we don't need to support VAX options forever.

**Comment by wlammen** [comment]
I'd rather see Metamath be split in a library servicing primitive operations, and on top of that a layer for the command line interface. But that's a lofty goal.

**Comment by digama0** [comment]
Definitely it would be nice to have a C interface that external programs can link against. If you combine that with the recent successful webassembly build of metamath.exe, we could get HTML pages served straight from the client, which would be super cool.

**Comment by benjub** [comment]
@wlammen's lofty goal and @digama0's (1) and (2) sound nice.  But does it require to remove TOOLS altogether ?  Why not keep it as is, maybe with a "use at your onw risks" warning if you want ?  I'm willing to break my muscle memory of course, especially to move toward more standard command line tools.  My remark about min.cmd also shows that some of the scripts may not work anymore, and would have to be updated (although they probably are in small number).

**Comment by digama0** [comment]
I hope to write a `submit.sh` soon along the lines of #34, which would take a tools file in the metamath format and evaluate it using the corresponding bash scripts. That should hopefully address the issues around compatibility with existing scripts like min.cmd, although I think that most tools mode scripts like min.cmd should be replaced by something less ad hoc eventually.
